examples/avro-cli.py

```diff
@@ -79,28 +94,31 @@ def on_delivery(err, msg, obj):
             obj.id, msg.topic(), msg.partition(), msg.offset()))
 
 
-def produce(topic, conf):
+def produce(topic, conf, avro_serializer):
     """
         Produce User records
     """
 
-    from confluent_kafka.avro import AvroProducer
-
-    producer = AvroProducer(conf, default_value_schema=record_schema)
+    producer = Producer(conf,
+                        key_serializer=StringSerializer('utf_8'),
+                        value_serializer=avro_serializer)
 
     print("Producing user records to topic {}. ^c to exit.".format(topic))
     while True:
         # Instantiate new User, populate fields, produce record, execute callbacks.
         record = User()
+        user_id = str(uuid4())
+
         try:
             record.name = input("Enter name: ")
             record.favorite_number = int(input("Enter favorite number: "))
             record.favorite_color = input("Enter favorite color: ")
 
             # The message passed to the delivery callback will already be serialized.
             # To aid in debugging we provide the original object to the delivery callback.
-            producer.produce(topic=topic, value=record.to_dict(),
-                             callback=lambda err, msg, obj=record: on_delivery(err, msg, obj))
+            producer.produce(topic=topic, key=user_id,
+                             value=record,
+                             on_delivery=lambda err, msg, obj=record: on_delivery(err, msg, obj))
```

Comment:
I'm sold on this approach now, but I think it would be really good if we could change the names: `Message.key()` -> `Message.key_bytes()` and `Message.value()` -> `Message.value_bytes()`, making the the API more immediately clear.

It seems likely the release with this PR in it will be called 2.0, so maybe that makes this more palatable. Would want to depreciate `key()` and `value()` rather than removing.

Reply:
There is nothing actually preventing you from constructing a dr cb that deserializes the message contents. I'll update the example to just do that

Reply:
My point is about making the API provided out of the box give the user what they most intuitively would expect. Not arguing you could override that with something that is better - point is you shouldn't need to.

Reply:
Perhaps, though this feels a bit heavy handed to me. This change does not have any effect on the overall behavior or result and yet it would break every existing message callback handler in use today. I think the costs outweigh the benefits on this one.

Reply:
you could consider making `key()`, `value()` throw an exception (probably 'not implemented yet') if the function was called with a non `byte[]` like arg (but return bytes if they were, so there is no breaking change). also add `key_bytes` and `value_bytes` which always return the bytes. in the future (i.e. probably plan on never, but leave the option open) you could add a config which would allow the original value to be passed through if enabled.

i think you should do this, but it's (nearly?) the last i'll say of it...

Reply:
>if the function was called with a non byte[] like arg 

What we could do is decorate any user supplied delivery callbacks on initialization and deserialize the contents prior to passing the message handle to the user's callback handler. We will want to be careful to catch any exceptions that may arise and set the msg error accordingly though. I'll wait for @edenhill to chime in before going that  route though.

Reply:
I went ahead and put a preliminary implementation in there.

Reply:
I strongly oppose having the delivery report perform deserialization: just pass the original object to the delivery report callback instead.
It seems like an anti-pattern to me to mix the serialized form/output with the internal program logic.

As for key() and value() changing meaning: yes, that is a problem, and it would be possible to change the semantics of key() and value() in the delivery report callback when using the SerializingProducer(), but then we're also always holding on to the passed object which has GC and lifetime issues which might not always be desired.

So I'd say we keep it simple for now and simply document that key() and value() are the serialized form iin the delivery report callback, and to use bound variables to pass the object as an opt-in if the original object is needed in the dr cb.